cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config

```diff
@@ -2,13 +2,45 @@
 password: fedora
 chpasswd: { expire: False }
 ssh_pwauth: yes
+write_files:
+    - path: /etc/modules-load.d/mlx5.conf
+      content: |
+        mlx5_core
+        mlx5_ib
+    - path: /etc/modules-load.d/mlx4.conf
+      content: |
+        mlx4_core
+        mlx4_ib
+    - path: /etc/modules-load.d/i40e.conf
+      content: |
+        i40e
+    - path: /etc/modules-load.d/igb.conf
+      content: |
+        igb
+    - path: /etc/systemd/system/wait-for-cloud-init.service
+      owner: root:root
+      permissions: '0755'
+      content: |
+          [Unit]
+          Description=Wait for cloud-init to complete
+          Before=qemu-guest-agent.service
+
+          [Service]
+          Type=oneshot
+          ExecStartPre=bash -c "while [ ! -f /var/lib/cloud/data/result.json ] && [ ! -f /etc/cloud/cloud-init.disabled ]; do sleep 1; done"
```

Comment:
very nice
maybe worth to wait for /var/lib/cloud/instance/boot-finished instead result.json ?
as this is the really last thing that occur i think ?

Reply:
hmm, but this service will run every boot right ?
it means that the first thing that need to be done is to clean the previous cloud-init indications (for resets)
or at least clean it on the first boot before the shutdown (in this file)

Reply:
Question - sorry for triple posts, just didnt want to EDIT, so it will jump
it means that one must config cloud-init or at least have the cloud-init disabled on the machine right (the 2nd is not possible via tests, chicken and egg maybe) ?
it means that it isn't perfect yet, and there is dependency, and burden to run manually stuff that otherwise we are not must to

Reply:
I have a better way already. Just wait a little. :)

Reply:
Ok, updated. I am now waiting until `cloud-final.service` is done. Then I do a blocking wait with `cloud-init status --wait` in a pre-check. Once the check passes, the `wait-for-cloud-init.service` is done. Since it is marked as something which has to run `Before` qemu-guest-agent the agent will be started afterwards.

This works in both cases, if cloud init is enabled or disabled:

 * If cloud-init gets enabled by the systemd generator: `wait-for-cloud-init` will be enabled and run, and qemu-guest-agent will wait for it
 * If cloud-init is disabled by  the systemd generator: Nothing will want `wait-for-cloud-init`. Since `wait-for-cloud-init` is not `wanted` by `qemu-guest-agent` it will be ignored in this case and `qemu-guest-agent` starts immediately.

Reply:
Nice !

Reply:
Nice !
once there will be the updated kubevirt PR we can review it please together

Just making sure please
`cloud-init status --wait` will work also after the 2nd reboot ? (first one after provision) ?
i mean the indication files are still there from the first boot, unless it doesn't use them

Reply:
Yes it always works.

Reply:
Looks good!

I have a question, will qemu-guest-agent start if we create a VM w/o cloud-init?

Reply:
Yes: If cloud-init is disabled by the systemd generator: Nothing will want wait-for-cloud-init. Since wait-for-cloud-init is not wanted by qemu-guest-agent it will be ignored in this case and qemu-guest-agent starts immediately. 

To rephrase it: For as long as no unit is enabled which `wants` `wait-for-cloud-init` it will not be pulled in, and therefore `qemu-guest-agent` will not wait for it. `qemu-guest-agent` is not even aware of it.

Reply:
FWIW I verified that `qemu-guest-agent` starts on a VM w/o cloud-config
``` 
[fedora@vmi-masquerade ~]$ sudo systemctl status wait-for-cloud-init.service
● wait-for-cloud-init.service - Wait for cloud-init to complete
     Loaded: loaded (/etc/systemd/system/wait-for-cloud-init.service; enabled; >
     Active: inactive (dead)
```
And for a VM **with** cloud-config that it does starts after cloud-init finished (using kubevirt example/vmi-masqurade.yaml)
```
...
[  168.941298] cloud-init[733]: ci-info: no authorized ssh keys fingerprints found for user fedora.
[  168.962179] cloud-init[733]: Cloud-init v. 19.4 finished at Mon, 26 Jul 2021 14:19:16 +0000. Datasource DataSourceNoCloud [seed=/dev/vdb][dsmode=net].  Up 168.85 seconds
[  OK  ] Finished Execute cloud user/final scripts.
[  OK  ] Reached target Cloud-init target.
         Starting Wait for cloud-init to complete...
[  OK  ] Finished Wait for cloud-init to complete.
[  OK  ] Started QEMU Guest Agent.
[  OK  ] Finished man-db-cache-update.service.
...
```
```
[fedora@vmi-masquerade2 ~]$ sudo systemctl status  wait-for-cloud-init.service
● wait-for-cloud-init.service - Wait for cloud-init to complete
     Loaded: loaded (/etc/systemd/system/wait-for-cloud-init.service; enabled; vendor preset: disabled)
     Active: active (exited) since Mon 2021-07-26 14:19:18 UTC; 8min ago
    Process: 3132 ExecStartPre=/usr/bin/cloud-init status --wait (code=exited, status=0/SUCCESS)
    Process: 4186 ExecStart=/usr/bin/true (code=exited, status=0/SUCCESS)
   Main PID: 4186 (code=exited, status=0/SUCCESS)
        CPU: 770ms

Jul 26 14:19:17 vmi-masquerade2 systemd[1]: Starting Wait for cloud-init to complete...
Jul 26 14:19:18 vmi-masquerade2 cloud-init[3132]: status: done
Jul 26 14:19:18 vmi-masquerade2 systemd[1]: Finished Wait for cloud-init to complete.
```

Reply:
Does it mean that now all the tests with cloud-init will have to wait for "agent connected condition" (even if they don't need/care about the ga)? Otherwise the test will be flaky since it will be able to connect to the vm before the cloud-init is ready.

Reply:
@AlonaKaplan yes, that was never different.